src/NHibernate/Linq/Visitors/LeftJoinRewriter.cs

```diff
@@ -28,26 +34,21 @@ public override void VisitAdditionalFromClause(AdditionalFromClause fromClause,
 
 			var mainFromClause = subQueryModel.MainFromClause;
 
-			var restrictions = subQueryModel.BodyClauses
-				.OfType<WhereClause>()
-				.Select(w => new NhWithClause(w.Predicate));
-
-			var join = new NhJoinClause(mainFromClause.ItemName,
-										mainFromClause.ItemType,
-										mainFromClause.FromExpression,
-										restrictions);
+			var join = new AdditionalFromClause(mainFromClause.ItemName, mainFromClause.ItemType, mainFromClause.FromExpression);
+			var restrictions = subQueryModel.BodyClauses.OfType<WhereClause>().ToList();
+			_parameters.AddLeftJoin(join, restrictions);
 
 			var innerSelectorMapping = new QuerySourceMapping();
 			innerSelectorMapping.AddMapping(fromClause, subQueryModel.SelectClause.Selector);
 
 			queryModel.TransformExpressions(ex => ReferenceReplacingExpressionTreeVisitor.ReplaceClauseReferences(ex, innerSelectorMapping, false));
 
 			queryModel.BodyClauses.RemoveAt(index);
-			queryModel.BodyClauses.Insert(index, @join);
-			InsertBodyClauses(subQueryModel.BodyClauses.Where(b => !(b is WhereClause)), queryModel, index + 1);
+			queryModel.BodyClauses.Insert(index, join);
+			InsertBodyClauses(subQueryModel.BodyClauses, queryModel, index + 1);
```

Comment:
No more filtering of where clauses: we need them in the model for ensuring expressions transformations are applied to them too.
If we were cloning them for building the restrictions list, then we would have to filter them out here then insert their clones.

They will not be added as HQL where because the query model visitor now check with visitor parameters each where clause for not being known as a with clause.